xbmc/cores/RetroPlayer/buffers/RenderBufferOpenGL.cpp

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (C) 2005-2018 Team Kodi
+ *  This file is part of Kodi - https://kodi.tv
+ *
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *  See LICENSES/README.md for more information.
+ */
+
+#include "RenderBufferOpenGL.h"
+#include "cores/RetroPlayer/rendering/RenderContext.h"
+
+using namespace KODI;
+using namespace RETRO;
+
+
+CRenderBufferOpenGL::CRenderBufferOpenGL(CRenderContext &context,
+                                         GLuint pixeltype,
+                                         GLuint internalformat,
+                                         GLuint pixelformat,
+                                         GLuint bpp) :
+  m_context(context),
+  m_pixeltype(pixeltype),
+  m_internalformat(internalformat),
+  m_pixelformat(pixelformat),
+  m_bpp(bpp)
+{
+}
+
+CRenderBufferOpenGL::~CRenderBufferOpenGL()
+{
+  DeleteTexture();
+}
+
+void CRenderBufferOpenGL::CreateTexture()
+{
+  glGenTextures(1, &m_textureId);
+
+  glBindTexture(m_textureTarget, m_textureId);
+
+  glTexImage2D(m_textureTarget, 0, m_internalformat, m_width, m_height, 0, m_pixelformat, m_pixeltype, NULL);
+
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+  glTexParameteri(m_textureTarget, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+
+  glBindTexture(m_textureTarget, 0);
+}
+
+bool CRenderBufferOpenGL::UploadTexture()
+{
+  if (!glIsTexture(m_textureId))
+    CreateTexture();
+
+  glBindTexture(m_textureTarget, m_textureId);
+
+  const int stride = GetFrameSize() / m_height;
+
+  glPixelStorei(GL_UNPACK_ALIGNMENT, m_bpp);
+
+  glPixelStorei(GL_UNPACK_ROW_LENGTH, stride / m_bpp);
+  glTexSubImage2D(m_textureTarget, 0, 0, 0, m_width, m_height, m_pixelformat, m_pixeltype, m_data.data());
```

Comment:
you should really consider using PBOs here. OpenGL rendering can't be done worse than with this code, by using glTexSubImage2D

Reply:
@FernetMenta I agree.
But would you mind me getting it in as is and improve the renderer in a second iteration?

Reply:
Sure. I would place a comment in. This is a big change and history is hard to follow.

Reply:
Yep, will do that.